source/mir/random/nonuniform.d

```diff
@@ -0,0 +1,246 @@
+/**
+Non-uniform distributions.
+
+License: $(LINK2 http://boost.org/LICENSE_1_0.txt, Boost License 1.0).
+
+Authors: Sebastian Wilzbach
+*/
+
+module mir.random.nonuniform;
+
+import std.traits : isNumeric;
+import std.stdio;
+
+/**
+Setup a _normal distribution sampler.
+
+Params:
+    T = floating-point type that should be sampled
+    R = unsigned UIntType of the random generator
+
+Returns:
+    A $(LREF Ziggurat) sampler that can be called to sample from the distribution.
+*/
+auto normal(T, R = uint)()
+{
+    import mir.internal.math : exp, log, sqrt;
+
+    auto pdf    = (T x) => cast(T) exp(T(-0.5) * x * x);
+    auto invPdf = (T x) => cast(T) sqrt(T(-2) * log(x));
+
+    // values from [Marsaglia00]
+    T rightEnd = 3.442619855899;
+
+    /// generate x for the last block (aka tail-block)
+    enum fallback = q{
+        T fallback(RNG)(bool isPositive, ref RNG gen) const
+        {
+            import std.random : uniform;
+            import std.math : log;
+            T x, y, u;
+            do
+            {
+                u = uniform!("[]", T, T)(0, 1, gen);
+                x = -log(u) / fs[$ - 1];
+                u = uniform!("[]", T, T)(0, 1, gen);
+                y = -log(u);
+            }
+            while(y + y < x * x);
+            return isPositive ? rightEnd + x : -rightEnd - x;
+        }
+    };
+
+    return Ziggurat!(T, fallback, R, true)(pdf, invPdf, 128, rightEnd, T(9.91256303526217e-3));
+}
+
+/**
+Setup an _exponential distribution sampler.
+
+Params:
+    T = floating-point type that should be sampled
+    R = unsigned UIntType of the random generator
+
+Returns:
+    A $(LREF Ziggurat) sampler that can be called to sample from the distribution.
+*/
+auto exponential(T, R = uint)()
+{
+    import mir.internal.math : exp, log;
+
+    auto pdf    = (T x) => cast(T) exp(-x);
+    auto invPdf = (T x) => cast(T) -log(x);
+
+    // values from [Marsaglia00]
+    enum fallback = q{
+        T fallback(RNG)(ref RNG gen) const
+        {
+            import std.random : uniform;
+            auto u = uniform!("[]", T, T)(0, 1, gen);
+            return 7.69711 - u;
+        }
+    };
+
+    return Ziggurat!(T, fallback, R, false)(pdf, invPdf, 256, T(7.697117470131487), T(3.949659822581572e-3));
```

Comment:
I put everything as template argument that is absolutely needed at compile time. 
Does it make a huge difference? - I thought in the end we can just use `enum z = Ziggurat!..`


On July 19, 2016 2:15:03 PM GMT+02:00, Joseph Wakeling <notifications@github.com> wrote:
>> +    import mir.internal.math : exp, log;
>> +
>> +    auto pdf    = (T x) => cast(T) exp(-x);
>> +    auto invPdf = (T x) => cast(T) -log(x);
>> +
>> +    // values from [Marsaglia00]
>> +    enum fallback = q{
>> +        T fallback(RNG)(ref RNG gen) const
>> +        {
>> +            import std.random : uniform;
>> +            auto u = uniform!("[]", T, T)(0, 1, gen);
>> +            return 7.69711 - u;
>> +        }
>> +    };
>> +
>> +    return Ziggurat!(T, fallback, R, false)(pdf, invPdf, 256,
>T(7.697117470131487), T(3.949659822581572e-3));
>
>While calculations using `pdf` and `invPdf` can only be done at runtime
>for now (because of the `exp` implementation issues), is there any
>reason why the actual lambdas can't be provided as template parameters?
>It would make for a more logical design, I think (and also be
>future-proof against the point when you get a CTFE'able `exp`).
>
>---
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/libmir/mir/pull/261/files/6a6565f321a80c67b95a7657bf55cc77989fa09c#r71325812